vfs/file.go

```diff
@@ -192,38 +202,96 @@ func CreateFileAndUpload(m *DocAttributes, fs afero.Fs, dbPrefix string, body io
 
 	var written int64
 	var md5Sum []byte
-	if written, md5Sum, err = copyOnFsAndCheckIntegrity(m, fs, pth, body); err != nil {
-		return
+	if written, md5Sum, err = copyOnFsAndCheckIntegrity(pth, doc.MD5Sum, doc.Executable, fs, body); err != nil {
+		return err
 	}
 
-	if attrs.Size < 0 {
-		attrs.Size = written
+	if doc.Size < 0 {
+		doc.Size = written
 	}
 
-	if attrs.MD5Sum == nil {
-		attrs.MD5Sum = md5Sum
+	if doc.MD5Sum == nil {
+		doc.MD5Sum = md5Sum
 	}
 
-	if attrs.Size != written {
-		err = ErrContentLengthMismatch
-		return
+	if doc.Size != written {
+		return ErrContentLengthMismatch
 	}
 
-	if err = couchdb.CreateDoc(dbPrefix, doc); err != nil {
-		return
+	doc.Path = pth
+
+	return couchdb.CreateDoc(dbPrefix, doc)
+}
+
+// ModifyFileContent overrides the content of a file onto the
+// filesystem.
+//
+// @TODO: make it more resilient to not lose data if the transfer
+// fails.
+func ModifyFileContent(oldDoc *FileDoc, newDoc *FileDoc, fs afero.Fs, dbPrefix string, body io.Reader) (err error) {
+	updateDate := time.Now()
+
+	pth := oldDoc.Path
+
+	defer func() {
+		if err != nil {
+			fs.Remove(pth)
+		}
+	}()
+
+	var written int64
+	var md5Sum []byte
+	if written, md5Sum, err = copyOnFsAndCheckIntegrity(pth, newDoc.MD5Sum, newDoc.Executable, fs, body); err != nil {
+		return err
 	}
 
-	return
+	if newDoc.Size < 0 {
+		newDoc.Size = written
+	}
+
+	if newDoc.MD5Sum == nil {
+		newDoc.MD5Sum = md5Sum
+	}
+
+	if newDoc.Size != written {
+		return ErrContentLengthMismatch
+	}
+
+	newDoc.Path = pth
+	newDoc.SetID(oldDoc.ID())
+	newDoc.SetRev(oldDoc.Rev())
+	newDoc.CreatedAt = oldDoc.CreatedAt
+	newDoc.UpdatedAt = updateDate
+
+	return couchdb.UpdateDoc(dbPrefix, newDoc)
 }
 
-func copyOnFsAndCheckIntegrity(m *DocAttributes, fs afero.Fs, pth string, r io.ReadCloser) (written int64, md5Sum []byte, err error) {
-	f, err := fs.Create(pth)
+func copyOnFsAndCheckIntegrity(pth string, givenMD5 []byte, executable bool, fs afero.Fs, r io.Reader) (written int64, md5Sum []byte, err error) {
+	// We want to write only (O_WRONLY), create the file if it does not
```

Comment:
This comment is at the wrong place (should be before L279). Another comment about "magic numbers" 0755 & 644 would be cool.